### ISSUE TITLE
feat: batched generation for Mamba/hybrid models (Nemotron-H)

### ISSUE THREAD

**Issue by haoxu0** [ml-explore/mlx-lm#1403]
## Problem

`batch_generate()` uses `BatchGenerator` which processes sequences via continuous batching — extract/filter/merge per finished sequence. For transformer models this works well since KV cache is read-only after prefill.

For **Mamba/SSM/hybrid models** (Nemotron-H, Jamba, Falcon Mamba), this is actually **slower** than sequential generation because:
1. Mamba's recurrent state must be updated each step — the "batch" overhead (extract/merge per token) exceeds any parallel benefit
2. The model's forward pass already supports batch dimension natively (`model(tokens_K×1)` works for K sequences simultaneously)

## What works instead

A simple batched decode loop that feeds all K sequences through one forward pass gives **3× speedup** on Nemotron-H (30B, 4-bit, M4 Max):

```python
# Prefill — all K sequences have same prompt
prompt_batch = mx.array([prompt_ids] * K)  # [K, prompt_len]
cache = make_prompt_cache(model)
logits = model(prompt_batch, cache=cache)  # processes all K at once
logits = logits[:, -1, :]  # [K, V]

# Decode — generate all K tokens per step in one forward pass
for _ in range(max_tokens):
    tokens = sample(logits)           # [K]
    logits = model(tokens.reshape(K, 1), cache=cache)
    logits = logits[:, -1, :]         # [K, V]
```

Benchmark (M4 Max, 64GB, Nemotron-H 30B 4-bit, K=8, 512 tokens):
- Sequential: 72s
- Batched forward: 24s → **3× speedup**
- `BatchGenerator`: 2.4× **slower** than sequential (overhead)

## Suggestion

For models with `make_cache()` returning `ArraysCache`/`MambaCache`, a specialized code path in `batch_generate()` (or a new `batched_decode()` function) that:
1. Duplicates the prompt K times 
2. Prefills once with batch dim
3. Decodes with `model(tokens_K×1, cache=cache)` — one forward call per step for all K sequences

This would complement `BatchGenerator` rather than replace it — `BatchGenerator` is better for transformers with heterogeneous prompts and variable-length sequences.

## Context

Using this for GRPO training on Apple Silicon where each step generates K=8-16 rollouts from the same prompt. The generation is 91% of training time (per mlx-tracer profiling).

Related: #1358 (logprobs from batch_generate)
